### PR TITLE
[RFC] Don't use GCC -O/-g flags with MSVC

### PR DESCRIPTION
**Pull request by sethjackson**
Cherry picked from #810. From equalsraf@d12c795.

### CODE REVIEW

**Comment by Pyrohh** [comment]
LGTM

**Comment by fwalch** [comment]
If MSVC doesn't support this anyway, the HAS_OG check could be moved into the if as well.

File: CMakeLists.txt [hunk 1]
```diff
@@ -156,14 +156,16 @@ else()
...
-else()
-  set(CMAKE_C_FLAGS_RELWITHDEBINFO "-O2 -g"
-    CACHE STRING "Flags used by the compiler during release builds with debug info." FORCE)
+if(NOT MSVC)
+  # Set custom build flags for RelWithDebInfo.
+  # -DNDEBUG purposely omitted because we want assertions.
+  if(HAS_OG_FLAG)
+    set(CMAKE_C_FLAGS_RELWITHDEBINFO "-Og -g"
+      CACHE STRING "Flags used by the compiler during release builds with debug info." FORCE)
+  else()
```
Comment (justinmk) [inline CMakeLists.txt]: Just change this to `elif`. Avoids the extra nesting.

Comment (sethjackson) [inline CMakeLists.txt]: Hmm like this?

```
if(NOT MSVC AND HAS_OG_FLAG)
  ...
elif(NOT MSVC)
  ...
end()
```

**Comment by sethjackson** [comment]
@fwalch I don't think there is anything specifically for debug optimization on MSVC.

The MSVC compiler flags are here: https://msdn.microsoft.com/en-us/library/19z1t1wy.aspx

We can however set the link flags for the type of runtime (dynamic/static) we are using and
equalsraf@cd5c23c does just that. However that flag addition can probably be done like this:

```
if(NOT MSVC AND HAS_OG_FLAG)
   ...
elseif(NOT MSVC)
   ...
else()
   set(CMAKE_C_FLAGS_RELWITHDEBINFO -MTd)
endif()
``

**Comment by fwalch** [comment]
@sethjackson Sorry, don't mind my comment, I didn't realize the check is only performed for GCC anyway.

I think this would be easier to read:

```
if(MSVC)
   ...
elseif(HAS_OG_FLAG)
  ...
else()
  ...
endif()
```

As for the MSVC flags, before overriding them with something we should verify the default. If it is `/MTd /DNDEBUG` for RelWithDebInfo, then changing it like this is fine (basically we want to use the "default RelWithDebInfo", but also enable assertions).

File: CMakeLists.txt [hunk 2]
```diff
@@ -158,10 +158,10 @@ endif()
 
 # Set custom build flags for RelWithDebInfo.
 # -DNDEBUG purposely omitted because we want assertions.
-if(HAS_OG_FLAG)
+if(NOT MSVC AND HAS_OG_FLAG)
```
Comment (justinmk) [inline CMakeLists.txt]: HAS_OG_FLAG will always be false here because of line 156, so I'll leave this out.

**Comment by justinmk** [comment]
Merged

**Pull request closed without merging**